src/io.rs

```diff
@@ -0,0 +1,95 @@
+/// be triggered by subsequent calls to one of the `_read_tls` methods.
+/// `userdata` is set to the value provided to `rustls_*_session_set_userdata`. In most
+/// cases that should be a struct that contains, at a minimum, a file descriptor.
+/// The buf and out_n pointers are borrowed and should not be retained across calls.
+pub type rustls_read_callback = Option<
+    unsafe extern "C" fn(
+        userdata: *mut c_void,
+        buf: *mut u8,
+        n: size_t,
+        out_n: *mut usize,
+    ) -> rustls_io_error,
+>;
+
+pub(crate) type ReadCallback = unsafe extern "C" fn(
+    userdata: *mut c_void,
+    buf: *mut u8,
+    n: size_t,
+    out_n: *mut usize,
+) -> rustls_io_error;
+
+pub(crate) struct CallbackReader {
+    pub callback: ReadCallback,
+    pub userdata: *mut c_void,
+}
+
+impl Read for CallbackReader {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        let mut out_n: usize = 0;
+        let cb = self.callback;
+        let result = unsafe { cb(self.userdata, buf.as_mut_ptr(), buf.len(), &mut out_n) };
+        match result.0 {
+            0 => Ok(out_n),
+            e => Err(Error::from_raw_os_error(e)),
+        }
+    }
+}
+
+/// A callback for rustls_server_session_write_tls or rustls_client_session_write_tls.
+/// An implementation of this callback should attempt to write the `n` bytes in buf
+/// to the network. If any bytes were written, the implementation should set out_n
+/// to the number of bytes written and return 0. If there was an error,
+/// (including EAGAIN or EWOULDBLOCK), the implementation should return `errno`.
+/// It's best to make one write attempt to the network per call. Additional write will
+/// be triggered by subsequent calls to one of the `_write_tls` methods.
+/// `userdata` is set to the value provided to `rustls_*_session_set_userdata`. In most
+/// cases that should be a struct that contains, at a minimum, a file descriptor.
+/// The buf and out_n pointers are borrowed and should not be retained across calls.
+pub type rustls_write_callback = Option<
+    unsafe extern "C" fn(
+        userdata: *mut c_void,
+        buf: *const u8,
+        n: size_t,
+        out_n: *mut usize,
+    ) -> rustls_io_error,
+>;
+
+pub(crate) type WriteCallback = unsafe extern "C" fn(
+    userdata: *mut c_void,
+    buf: *const u8,
+    n: size_t,
+    out_n: *mut usize,
+) -> rustls_io_error;
+
+pub(crate) struct CallbackWriter {
+    pub callback: WriteCallback,
+    pub userdata: *mut c_void,
+}
+
+impl Write for CallbackWriter {
+    fn write(&mut self, buf: &[u8]) -> Result<usize> {
+        let mut out_n: usize = 0;
+        let cb = self.callback;
+        let result = unsafe { cb(self.userdata, buf.as_ptr(), buf.len(), &mut out_n) };
+        match result.0 {
+            0 => Ok(out_n),
+            e => Err(Error::from_raw_os_error(e)),
+        }
+    }
+
+    fn flush(&mut self) -> Result<()> {
+        Ok(())
```

Comment:
Might we want to instruct the client to flush whatever buffers they may be using? I think that might require a separate flush callback which is kind of a hassle but I worry about cases where `[c]rustls` assume that some bytes have gone out over the network because `std::io::Write::flush()` succeeded but they're actually still buffered somewhere.

Reply:
As a practical matter, rustls' `write_tls` doesn't call `.flush()` on the `Write` its passed. So I'd rather not introduce the second callback, since it will never be called. Note that if the user wants the bytes flushed, they can call their OS' flush immediately after `write_tls()`.